model/xgboost/tests/test_model.py

```diff
@@ -0,0 +1,91 @@
+from dffml.feature.feature import Feature, Features
+from dffml.source.source import Sources
+from dffml.util.entrypoint import entrypoint
+from dffml.record import Record
+from dffml.util.asynctestcase import AsyncTestCase
+
+from dffml_model_xgboost.xdgregressor import (
+    XDGRegressorModel,
+    XDGRegressorModelConfig,
+)
+
+
+class TestXDGRegressor(AsyncTestCase):
+    @classmethod
+    def setUpClass(cls):
+        # Create a temporary directory to store the trained model
+        cls.model_dir = tempfile.TemporaryDirectory()
+        # Create an instance of the model
+        cls.model = XDGRegressorModel(
+            XDGRegressorModelConfig(
+                features=Features(
+                    Feature("Feature1", float, 1), Feature("Feature2")
+                ),
+                predict=Feature("Target", float, 1),
+                directory=cls.model_dir.name,
+            )
+        )
+        # Generating data f(x1,x2) = 2*x1 + 3*x2
+        _n_data = 2000
+        _temp_data = np.random.rand(2, _n_data)
+        cls.records = [
+            Record(
+                "x" + str(random.random()),
+                data={
+                    "features": {
+                        "Feature1": float(_temp_data[0][i]),
+                        "Feature2": float(_temp_data[1][i]),
+                        "Target": 2 * _temp_data[0][i] + 3 * _temp_data[1][i],
+                    }
+                },
+            )
+            for i in range(0, _n_data)
+        ]
+
+        cls.trainingsource = Sources(
+            MemorySource(MemorySourceConfig(records=cls.records[:1800]))
+        )
+        cls.testsource = Sources(
+            MemorySource(MemorySourceConfig(records=cls.records[1800:]))
+        )
+
+    @classmethod
+    def tearDownClass(cls):
+        # Remove the temporary directory where the model was stored to cleanup
+        cls.model_dir.cleanup()
+
+    async def test_00_train(self):
+        # Train the model on the training data
+        await train(self.model, self.trainingsource)
+
+    async def test_01_accuracy(self):
+        # Use the test data to assess the model's accuracy
+        res = await accuracy(self.model, self.testsource)
+        # Ensure the accuracy is above 80%
+        self.assertTrue(0.8 <= res < 1.0)
+
+    async def test_02_predict(self):
+        # Get the prediction for each piece of test data
+        async for i, features, prediction in predict(
+            self.model, self.testsource
+        ):
+            # Grab the correct value
+            correct = features["Target"]
+            # Grab the predicted value
+            prediction = prediction["Target"]["value"]
+            # Check that the prediction is within 30% error of the actual value
+            error = abs((prediction - correct) / correct)
+
+            acceptable = 0.3
+            # Sometimes causes an issue when only one data point anomalously has high error
+            self.assertLess(error, acceptable)
```

Comment:
We should be sure to test the example as well. You could add a `test_example` function that uses `subprocess.check_call` and `os.path.join(os.dirname(__file__), "..", "examples", "diabetesregression.py")` to run the file

Reply:
I made this changed in my new commit. However, I am getting an odd error when I run the test that I don't understand. It seems to be failing on the import of dffml in the example. Here is the traceback:`Traceback (most recent call last):
  File "c:\users\olive\onedrive - university of cambridge\documents\github\dffml\model\xgboost\tests\..\examples\diabetesregression.py", line 6, in <module>
    from dffml import Feature, Features
  File "c:\users\olive\onedrive - university of cambridge\documents\github\dffml\dffml\__init__.py", line 69, in <module>
    for import_name, module in modules(root, package_name, skip=skip):
  File "c:\users\olive\onedrive - university of cambridge\documents\github\dffml\dffml\__init__.py", line 30, in modules
    yield import_name, importlib.import_module(import_name)
  File "C:\Users\olive\Anaconda2\envs\py36\lib\importlib\__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "c:\users\olive\onedrive - university of cambridge\documents\github\dffml\dffml\high_level.py", line 10, in <module>
    from .df.types import DataFlow, Input
  File "c:\users\olive\onedrive - university of cambridge\documents\github\dffml\dffml\df\types.py", line 403, in <module>
    @dataclass
  File "c:\users\olive\anaconda2\envs\py36\lib\site-packages\dataclasses.py", line 958, in dataclass
    return wrap(_cls)
  File "c:\users\olive\anaconda2\envs\py36\lib\site-packages\dataclasses.py", line 950, in wrap
    return _process_class(cls, init, repr, eq, order, unsafe_hash, frozen)
  File "c:\users\olive\anaconda2\envs\py36\lib\site-packages\dataclasses.py", line 801, in _process_class
    for name, type in cls_annotations.items()]
  File "c:\users\olive\anaconda2\envs\py36\lib\site-packages\dataclasses.py", line 801, in <listcomp>
    for name, type in cls_annotations.items()]
  File "c:\users\olive\anaconda2\envs\py36\lib\site-packages\dataclasses.py", line 659, in _get_field
    if (_is_classvar(a_type, typing)
  File "c:\users\olive\anaconda2\envs\py36\lib\site-packages\dataclasses.py", line 550, in _is_classvar
    return type(a_type) is typing._ClassVar
AttributeError: module 'typing' has no attribute '_ClassVar'
ERROR

======================================================================
ERROR: test_03_example (tests.test_model.TestXDGRegressor)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "c:\users\olive\onedrive - university of cambridge\documents\github\dffml\dffml\util\asynctestcase.py", line 69, in run_it
    result = self.loop.run_until_complete(coro(*args, **kwargs))
  File "C:\Users\olive\Anaconda2\envs\py36\lib\asyncio\base_events.py", line 587, in run_until_complete
    return future.result()
  File "c:\users\olive\onedrive - university of cambridge\documents\github\dffml\model\xgboost\tests\test_model.py", line 105, in test_03_example
    subprocess.check_call([sys.executable, filepath])
  File "C:\Users\olive\Anaconda2\envs\py36\lib\subprocess.py", line 363, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['C:\\Users\\olive\\Anaconda2\\envs\\py36\\python.exe', 'c:\\users\\olive\\onedrive - university of cambridge\\documents\\github\\dffml\\model\\xgboost\\tests\\..\\examples\\diabetesregression.py']' returned non-zero exit status 1.`

Reply:
It seems that this problem doesn't appear in the CI tests, so maybe it is to do with my local enviroment

Reply:
Ah it looks like you anaconda environment mentions `/py36/` in the traceback path. We only support Python 3.7 and 3.8 (due to some of these dataclasses features and asyncio features)